File: growmore-frontend/types/portfolio.ts
export interface Portfolio {
  id: string;
  user_id: string;
  name: string;
  description?: string;
  is_default: boolean;
  created_at: string;
  updated_at: string;
}

export type AssetType = 'stock' | 'commodity' | 'bank_product' | 'other';

export interface Holding {
  id: string;
  portfolio_id: string;
  asset_type: AssetType;
  asset_id: string;
  symbol?: string;
  name: string;
  quantity: number;
  avg_cost: number;
  current_price: number;
  invested_value: number;
  current_value: number;
  profit_loss: number;
  profit_loss_percent: number;
  allocation_percent: number;
  created_at: string;
  updated_at: string;
}

export interface PortfolioWithHoldings extends Portfolio {
  holdings: Holding[];
  total_value: number;
  total_invested: number;
  total_profit_loss: number;
  total_profit_loss_percent: number;
  today_change: number;
  today_change_percent: number;
}

export type TransactionType = 'buy' | 'sell' | 'dividend' | 'other';

export interface Transaction {
  id: string;
  portfolio_id: string;
  holding_id?: string;
  transaction_type: TransactionType;
  asset_type: AssetType;
  asset_id: string;
  symbol?: string;
  name: string;
  quantity: number;
  price: number;
  total_amount: number;
  fees?: number;
  notes?: string;
  transaction_date: string;
  created_at: string;
}

export interface PortfolioPerformance {
  portfolio_id: string;
  total_value: number;
  total_invested: number;
  total_return: number;
  total_return_percent: number;
  today_change: number;
  today_change_percent: number;
  best_performer?: {
    symbol: string;
    name: string;
    return_percent: number;
  };
  worst_performer?: {
    symbol: string;
    name: string;
    return_percent: number;
  };
  allocation_by_type: {
    asset_type: AssetType;
    value: number;
    percent: number;
  }[];
  allocation_by_sector?: {
    sector: string;
    value: number;
    percent: number;
  }[];
  historical_performance: {
    date: string;
    value: number;
  }[];
}

export interface HoldingCreate {
  asset_type: AssetType;
  asset_id: string;
  symbol?: string;
  name: string;
  quantity: number;
  avg_cost: number;
}

export interface HoldingUpdate {
  quantity?: number;
  avg_cost?: number;
}

export interface TransactionCreate {
  holding_id?: string;
  transaction_type: TransactionType;
  asset_type: AssetType;
  asset_id: string;
  symbol?: string;
  name: string;
  quantity: number;
  price: number;
  fees?: number;
  notes?: string;
  transaction_date: string;
}

File: growmore-backend/app/models/portfolio.py
from datetime import datetime
from decimal import Decimal
from typing import Optional
from uuid import UUID

from pydantic import BaseModel


class Portfolio(BaseModel):
    id: UUID
    user_id: UUID
    name: str
    description: Optional[str] = None
    is_default: bool = False
    total_invested: Decimal = Decimal("0")
    current_value: Decimal = Decimal("0")
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PortfolioHolding(BaseModel):
    id: UUID
    portfolio_id: UUID
    holding_type: str
    holding_id: UUID
    quantity: Decimal
    avg_buy_price: Decimal
    total_invested: Decimal
    current_value: Optional[Decimal] = None
    profit_loss: Optional[Decimal] = None
    profit_loss_percentage: Optional[Decimal] = None
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PortfolioTransaction(BaseModel):
    id: UUID
    portfolio_id: UUID
    holding_type: str
    holding_id: UUID
    transaction_type: str
    quantity: Decimal
    price: Decimal
    total_amount: Decimal
    fees: Decimal = Decimal("0")
    notes: Optional[str] = None
    transaction_date: datetime
    created_at: datetime

    class Config:
        from_attributes = True
